<a href="https://colab.research.google.com/github/optimalMachine/EliteDataScienceCourse/blob/master/5_Workbook_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### COMPANION WORKBOOK

# Pandas

To make the most out this program, we strongly recommend you to:
1. First practice writing and implementing all of the code from Coding Section of the online lesson.
2. Then, freely experiment with and explore any interesting or confusing concepts. Simply insert new code cells and then use the help of Google and official documentation.
3. Finally, tackle all of the exercises at the end. They will help you tie everything together and **learn in context.**

#### <span style="color:#555">LESSON CODE SANDBOX</span>

Use this space to practice writing and implementing all of the code from Coding Section of the online lesson. Insert new code cells as needed, and feel free to write notes to yourself in Markdown.

## I. Pandas DataFrames are like spreadsheets.

In [36]:
import pandas as pd

In [37]:
example_dataframe = pd.DataFrame({'column_1':[5,4,3],'column_2':['a','b','c']})
example_dataframe

column_1 column_2
0         5        a
1         4        b
2         3        c

In [38]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [39]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Python for Data Science/project_files/iris.csv')
print (type(df))

<class 'pandas.core.frame.DataFrame'>


In [40]:
df.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [41]:
print (df.shape)

(150, 5)


In [42]:
print (len(df))
print (df.shape[0])

150
150


In [43]:
print (df.min())

sepal_length       4.3
sepal_width        2.0
petal_length       1.0
petal_width        0.1
species         setosa
dtype: object


In [44]:
print (df.max())

sepal_length          7.9
sepal_width           4.4
petal_length          6.9
petal_width           2.5
species         virginica
dtype: object


In [45]:
df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

## II. Pandas Series are like single columns.

In [46]:
integer_series = pd.Series([0,1,2,3,4])
print (integer_series)
print (type(integer_series))

0    0
1    1
2    2
3    3
4    4
dtype: int64
<class 'pandas.core.series.Series'>


In [47]:
print (type(df.petal_length))

<class 'pandas.core.series.Series'>


In [48]:
print (type(df['petal_length']))

<class 'pandas.core.series.Series'>


In [49]:
print (df.petal_length.head())
print ('The minimum petal length is', df.petal_length.min())
print ('The maximum petal length is', df.petal_length.max())

0    1.4
1    1.4
2    1.3
3    1.5
4    1.4
Name: petal_length, dtype: float64
The minimum petal length is 1.0
The maximum petal length is 6.9


In [50]:
print (df.species.unique())

['setosa' 'versicolor' 'virginica']


In [51]:
print (df.petal_length.median())
import numpy as np
print (np.median(df.petal_length))

4.35
4.35


In [52]:
df['petal_area'] = df.petal_width * df.petal_length
df['petal_area'].head()

0    0.28
1    0.28
2    0.26
3    0.30
4    0.28
Name: petal_area, dtype: float64

## III. Boolean masks allow you to filter data.

In [53]:
list_mask = [True,True,True,True]
series_mask = ([True,True,False,False])

In [54]:
print (series_mask)

[True, True, False, False]


In [71]:
example_series = pd.Series([10,5,-3,2])
print (example_series)

0    10
1     5
2    -3
3     2
dtype: int64


In [72]:
series_mask = example_series > 0
print (series_mask)

0     True
1     True
2    False
3     True
dtype: bool


In [73]:
example_series[series_mask]

0    10
1     5
3     2
dtype: int64

## IV. Groupbys allow you to segment and aggregate data.

#### <span style="color:#555">EXERCISES</span>

Complete each of the following exercises.

First, remember to import Pandas and read in the iris dataset.

In [56]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [57]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Python for Data Science/project_files/iris.csv')

## <span style="color:RoyalBlue">Exercise 5.1 - Flowers for Casey</span>

In the previous lesson, we discovered that the first local business we need to visit in Park Royal is a local **flower shop.** When we arrive, we find out that the owner, Casey, urgently needs our help!

They just received a new shipment of **iris flowers**, but they've never stocked these flowers before. Casey asks us to share what we know about these new flowers. Luckily, we have the *iris dataset* to learn more about them ourselves!

In the previous lesson, we also introduced the idea of using a **toy problem** to reduce the size of a problem and break down the concepts. While 150 observations is not exactly "big data," it's still too large to fit on our screen at once. So let's use another toy problem to practice the concepts in this lesson.

#### A.) First, create a new DataFrame called <code style="color:steelblue">toy_df</code>. It should contain the first 5 rows plus the last 5 rows from our original Iris dataset.
* **Tip:** You already have a <code style="color:steelblue">.head()</code>, but what about a <code style="color:steelblue">.tail()</code>?
* **Tip:** <code style="color:steelblue">pd.concat()</code> is your [friend](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html).

In [58]:
# Create toy_df


#### B.) Next, display <code style="color:steelblue">toy_df</code>.
* After all, it will only be 10 rows.
* In <code style="color:steelblue">toy_df</code>, you should have data from 2 different species of flower. Which are they?

In [59]:
# Display toy_df


<strong style="color:RoyalBlue">Expected output:</strong>

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4.9</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.3</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.6</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>3.6</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>145</th>
      <td>6.7</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.3</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>146</th>
      <td>6.3</td>
      <td>2.5</td>
      <td>5.0</td>
      <td>1.9</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>147</th>
      <td>6.5</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.0</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>148</th>
      <td>6.2</td>
      <td>3.4</td>
      <td>5.4</td>
      <td>2.3</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>149</th>
      <td>5.9</td>
      <td>3.0</td>
      <td>5.1</td>
      <td>1.8</td>
      <td>virginica</td>
    </tr>
  </tbody>
</table>

#### C.) Next, display a summary table for <code style="color:steelblue">toy_df</code>.
* It should have the mean, standard deviation, and quartiles for each of the columns.
* Since <code style="color:steelblue">toy_df</code> is only 10 rows, you can manually check the **mins** and **maxes**. Are they correct?

In [60]:
# Describe toy_df


<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>10.000000</td>
      <td>10.000000</td>
      <td>10.000000</td>
      <td>10.000000</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>5.590000</td>
      <td>3.130000</td>
      <td>3.290000</td>
      <td>1.130000</td>
    </tr>
    <tr>
      <th>std</th>
      <td>0.807534</td>
      <td>0.316403</td>
      <td>1.995244</td>
      <td>0.992248</td>
    </tr>
    <tr>
      <th>min</th>
      <td>4.600000</td>
      <td>2.500000</td>
      <td>1.300000</td>
      <td>0.200000</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>4.925000</td>
      <td>3.000000</td>
      <td>1.400000</td>
      <td>0.200000</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>5.500000</td>
      <td>3.050000</td>
      <td>3.250000</td>
      <td>1.000000</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>6.275000</td>
      <td>3.350000</td>
      <td>5.175000</td>
      <td>1.975000</td>
    </tr>
    <tr>
      <th>max</th>
      <td>6.700000</td>
      <td>3.600000</td>
      <td>5.400000</td>
      <td>2.300000</td>
    </tr>
  </tbody>
</table>

## <span style="color:RoyalBlue">Exercise 5.2 - Basic Feature Engineering</span>

Elementwise operations are very useful in machine learning, especially for **feature engineering.** Feature engineering is the process of creating new features (model input variables) from existing ones, and its one of the best ways for data scientists to add value.

In the Iris dataset, we have petal width and length, but what if we wanted to know petal area? Well, we can create a new <code style="color:steelblue">petal_area</code> feature (yes, the petals are not perfect rectangles, but that's fine).

#### A.) First, display the two columns of <code style="color:steelblue">petal_width</code> and <code style="color:steelblue">petal_length</code> in <code style="color:steelblue">toy_df</code>.
* **Tip:** You can index a DataFrame using a list of column names too, like so:


<pre style="color:steelblue">df[['column_1', 'column_2']]</pre>

In [61]:
# Display petal_width and petal_length


<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>petal_width</th>
      <th>petal_length</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.2</td>
      <td>1.4</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.2</td>
      <td>1.4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.2</td>
      <td>1.3</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.2</td>
      <td>1.5</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.2</td>
      <td>1.4</td>
    </tr>
    <tr>
      <th>145</th>
      <td>2.3</td>
      <td>5.2</td>
    </tr>
    <tr>
      <th>146</th>
      <td>1.9</td>
      <td>5.0</td>
    </tr>
    <tr>
      <th>147</th>
      <td>2.0</td>
      <td>5.2</td>
    </tr>
    <tr>
      <th>148</th>
      <td>2.3</td>
      <td>5.4</td>
    </tr>
    <tr>
      <th>149</th>
      <td>1.8</td>
      <td>5.1</td>
    </tr>
  </tbody>
</table>

#### B.) Next, create a new <code style="color:steelblue">petal_area</code> feature in <code style="color:steelblue">toy_df</code>.
* Multiply the <code style="color:steelblue">petal_width</code> column by the <code style="color:steelblue">petal_length</code> column.
* Display <code style="color:steelblue">toy_df</code> after creating the new feature.
* Are the values for <code style="color:steelblue">petal_area</code> correct? Manually spot check a few of them just to make sure.

In [62]:
# Create a new petal_area column


# Display toy_df



<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
      <th>petal_area</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4.9</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.3</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.26</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.6</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.30</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>3.6</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>145</th>
      <td>6.7</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>11.96</td>
    </tr>
    <tr>
      <th>146</th>
      <td>6.3</td>
      <td>2.5</td>
      <td>5.0</td>
      <td>1.9</td>
      <td>virginica</td>
      <td>9.50</td>
    </tr>
    <tr>
      <th>147</th>
      <td>6.5</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.0</td>
      <td>virginica</td>
      <td>10.40</td>
    </tr>
    <tr>
      <th>148</th>
      <td>6.2</td>
      <td>3.4</td>
      <td>5.4</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>12.42</td>
    </tr>
    <tr>
      <th>149</th>
      <td>5.9</td>
      <td>3.0</td>
      <td>5.1</td>
      <td>1.8</td>
      <td>virginica</td>
      <td>9.18</td>
    </tr>
  </tbody>
</table>

### C.) Finally, what do we now know about Iris flowers?

By creating a <code style="color:steelblue">petal_area</code> feature, it's now much easier to see that virginica flowers have significantly larger petals than setosa flowers do! Often, by creating new features, you can learn more about the data (and improve your machine learning models). As you might guess, a machine learning algorithm and take these features and tease out the underlying patterns in a much more precise and comprehensive way.

## <span style="color:RoyalBlue">Exercise 5.3 - Mask On Mask Off</span>

Now, let's use the <code style="color:steelblue">toy_df</code> to really drive home the concept of **boolean masks**, because we'll be using these very frequently in data science.

For example, assume we wanted to display observations where <code style="color:steelblue">petal_area > 10</code> and <code style="color:steelblue">sepal_width > 3</code>. How could we do so?

#### A.) First, display <code style="color:steelblue">toy_df</code> again just to have it in front of you.

In [63]:
# Display toy_df


#### B.) Take a look at the DataFrame and manually count the number that satisfy our conditions.
* How many observations have <code style="color:steelblue">petal_area > 10</code>?
* How many observations have <code style="color:steelblue">sepal_width > 3</code>?
* How many satisfy both conditions?

*In the toy_df, there are:*
* **3** observations with <code style="color:steelblue">petal_area > 10</code> (indices 145, 147, 148)
* **5** observations with <code style="color:steelblue">sepal_width > 3</code> (indices 0,2,3,4,148)
* **1** observation that satifies both conditions (index 148)

Great. Now we'll see what's going on under the hood when we use our boolean masks.

#### C.) Create a boolean mask for <code style="color:steelblue">petal_area > 10</code>.
* Name it <code style="color:steelblue">petal_area_mask</code>.
* Display the mask after you create it.
* Does the result make sense?

In [64]:
# Mask for petal_area > 10


# Display petal_area_mask


<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
0      False
1      False
2      False
3      False
4      False
145     True
146    False
147     True
148     True
149    False
Name: petal_area, dtype: bool
</pre>

#### D.) Next, create a boolean mask for <code style="color:steelblue">sepal_width > 3</code>.
* Name it <code style="color:steelblue">sepal_width_mask</code>.
* Display the mask after you create it.
* Does the result make sense?

In [65]:
# Mask for sepal_width > 3


# Display sepal_width_mask


<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
0       True
1      False
2       True
3       True
4       True
145    False
146    False
147    False
148     True
149    False
Name: sepal_width, dtype: bool
</pre>

#### E.) Next, display the two masks combined using the <code style="color:steelblue">&</code> operator.
* Note how their combination results in another boolean mask!

In [66]:
# Display both masks, combined


<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
0      False
1      False
2      False
3      False
4      False
145    False
146    False
147    False
148     True
149    False
dtype: bool
</pre>

#### F.) Finally, select the observation(s) from <code style="color:steelblue">toy_df</code> where both conditions are met.
* Do you get the result you expected based on (B)?

In [67]:
# Index with both masks


## <span style="color:RoyalBlue">Exercise 5.4 - Iris Insights</span>

With the help of **groupbys**, we're almost ready to return to Casey and share what we've learned about Iris flowers. But before we do, let's just bring back our <code style="color:steelblue">toy_df</code> for one last hoorah, just to make sure we know what's going on under the hood.

Let's calculate the median <code style="color:steelblue">petal_area</code> for each species. Since <code style="color:steelblue">toy_df</code> is small, we can do this manually as well and check to make sure the values are correct.


#### A.) First, let's manually calculate the median <code style="color:steelblue">petal_area</code> for the virginica flowers in our <code style="steelblue">toy_df</code>.
* Display all observations of the virginica species.
* Sort them by <code style="color:steelblue">petal_area</code> in ascending order.
* **Tip:** Check out the <code style="color:steelblue">.sort_values()</code> function ([documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)).

In [68]:
# Display all 'virginica' species, sorted by petal_area


<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
      <th>petal_area</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>149</th>
      <td>5.9</td>
      <td>3.0</td>
      <td>5.1</td>
      <td>1.8</td>
      <td>virginica</td>
      <td>9.18</td>
    </tr>
    <tr>
      <th>146</th>
      <td>6.3</td>
      <td>2.5</td>
      <td>5.0</td>
      <td>1.9</td>
      <td>virginica</td>
      <td>9.50</td>
    </tr>
    <tr>
      <th>147</th>
      <td>6.5</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.0</td>
      <td>virginica</td>
      <td>10.40</td>
    </tr>
    <tr>
      <th>145</th>
      <td>6.7</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>11.96</td>
    </tr>
    <tr>
      <th>148</th>
      <td>6.2</td>
      <td>3.4</td>
      <td>5.4</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>12.42</td>
    </tr>
  </tbody>
</table>

Based on the output above, what's median <code style="color:steelblue">petal_area</code> for the virginica species?

#### B.) Next, let's manually calculate the median <code style="color:steelblue">petal_area</code> for the setosa flowers in our <code style="steelblue">toy_df</code>.
* Display all observations of the setosa species.
* Sort them by <code style="color:steelblue">petal_area</code> in ascending order.
* Based on the output, what's median <code style="color:steelblue">petal_area</code> for the setosa species?

In [69]:
# Display all 'setosa' species


<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
      <th>petal_area</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.3</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.26</td>
    </tr>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4.9</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>3.6</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.6</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.30</td>
    </tr>
  </tbody>
</table>

#### C.) Finally let's calculate the median values using a <code style="color:steelblue">.groupby()</code>.
* Do you get the same result?

In [70]:
# Median petal_area in toy_df


<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
species
setosa        0.28
virginica    10.40
Name: petal_area, dtype: float64
</pre>

#### Congratulations... You've completed the training mission!

In this lesson, you explored the Iris dataset using Pandas. After arming yourself with Pandas, you've successfully completed the rest of the training mission!*